### PR TITLE
Parse xml and htm[l]? documents

### LINKED ISSUE #30: Separately parse xml and html documents
See #18

### PR DESCRIPTION
**Pull request by adam3smith**
Closes #30 
ideas for doing this more elegantly welcome, of course

### CODE REVIEW

**Comment by adam3smith** [comment]
Closes #12

**Comment by adam3smith** [comment]
@greebie if you'd have time to review this? If code is OK, I'd add to docs and then I think we have this in good enough shape to start promoting, even if we aren't quite ready for CRAN yet.

File: R/archivr.R [hunk 1]
```diff
@@ -298,13 +299,14 @@ view_archiv.fromText <- function (fp, method="wayback") {
 #' archiv.fromUrl("https://www-cs-faculty.stanford.edu/~knuth/retd.html", method="perma_cc")
 #' }
 archiv.fromUrl <- function (url, method="wayback") {
-  return(archiv(extract_urls_from_webpage(url), method))
+  return(archiv(extract_urls_from_webpage(url, except), method))
```
Comment (greebie) [inline R/archivr.R:318]: Need to include except in function call.

**Review by greebie — CHANGES_REQUESTED** [review]
Just a minor issue that broke Travis and Appveyor (see review).

File: R/archivr.R [hunk 2]
```diff
@@ -259,7 +260,7 @@ view_archiv <- function (lst, method="wayback") {
 #' @return a dataframe containing the url, status, availability,
 #'   archived url(s) and timestamp(s)
 #' @examples
-#' view_archiv.fromUrl("https://www-cs-faculty.stanford.edu/~knuth/retd.html", method="both")
+#' checkArchiveStatus <- view_archiv.fromUrl("https://www-cs-faculty.stanford.edu/~knuth/retd.html", method="both")
```
Comment (greebie) [inline R/archivr.R]: Checkreview complained about the example line being over 100 characters.  I suggest:

``` 
#' checkArchiveStatus <- view_archiv.fromUrl(
#'     "https (..)",
#'     method="both")

File: R/archivr.R [hunk 3]
```diff
@@ -273,7 +274,7 @@ view_archiv.fromUrl <- function (url, method="wayback") {
 #'   archived url(s) and timestamp(s)
 #' @examples
 #' \dontrun{\
-#' view_archiv.fromText("testfile.docx", method="both")
+#' checkArchiveStatus <- view_archiv.fromText("testfile.docx", method="both")
```
Comment (greebie) [inline R/archivr.R:286]: Same.

File: R/archivr.R [hunk 4]
```diff
@@ -284,66 +285,68 @@ view_archiv.fromText <- function (fp, method="wayback") {
...
 #' @param url The url to extract links from.
 #' @param method Either "wayback," "perma_cc" or "both".
+#' @param except A regular expression for URLs to exclude from extraction
 #' @export
 #' @return a dataframe containing the url, status, availability,
 #'   archived url(s) and timestamp(s)
 #' @examples
 #' # Wayback
-#' archiv.fromUrl("https://www-cs-faculty.stanford.edu/~knuth/retd.html")
+#' archivedURLs <- archiv.fromUrl("https://www-cs-faculty.stanford.edu/~knuth/retd.html", except="validator\\.w3\\.org")
```
Comment (greebie) [inline R/archivr.R]: Same.

File: R/archivr.R [hunk 5]
```diff
@@ -284,66 +285,68 @@ view_archiv.fromText <- function (fp, method="wayback") {
...
 #' \dontrun{
 #' set_api_key("API KEY")
 #' set_folder_id("42")
-#' archiv.fromUrl("https://www-cs-faculty.stanford.edu/~knuth/retd.html", method="perma_cc")
+#' archivedURLs <- archiv.fromUrl("https://www-cs-faculty.stanford.edu/~knuth/retd.html", method="perma_cc")
 #' }
-archiv.fromUrl <- function (url, method="wayback") {
-  return(archiv(extract_urls_from_webpage(url), method))
+archiv.fromUrl <- function (url, method="wayback", except = NULL) {
+  return(archiv(extract_urls_from_webpage(url, except), method))
```
Comment (greebie) [inline R/archivr.R:318]: Same.

File: R/archivr.R [hunk 6]
```diff
@@ -284,66 +285,68 @@ view_archiv.fromText <- function (fp, method="wayback") {
...
-#'     was crawled.
+#' @return a list containing
+#'   the original url.
+#'   the http status
+#'   TRUE if successful or FALSE
+#'   the archived url.
+#'   the last time the url was crawled.
 #' @examples
-#' from_wayback("https://www-cs-faculty.stanford.edu/~knuth/retd.html")
+#' checkStatus <- from_wayback("https://www-cs-faculty.stanford.edu/~knuth/retd.html")
```
Comment (greebie) [inline R/archivr.R]: Same.

**Comment by greebie** [comment]
Travis did not work this time. Seems like a problem with from_wayback?

```
  > ### ** Examples
  > 
  > urls <- c("https://qdr.syr.edu", "https://cran.r-project.org/", "https://apsa.net")
  > 
  > # archive in Wayback machine
  > archiv(urls)
  [1] "Received a NULL value from archived snapshots from Wayback for https://qdr.syr.edu"
  [1] "Discovered an error in saving the url. Received http status 502. Perhaps try again at another time."
  Error in names(x) <- value : 
    'names' attribute [5] must be the same length as the vector [4]
  Calls: archiv -> colnames<-
  Execution halted
```

File: R/archivr.R (hunk 6 above)
Comment (adam3smith) [reply R/archivr.R]: Will fix, thanks, didn't see these comments before the latest commit.

**Comment by adam3smith** [comment]
@greebie -- we're warning free, thanks! Let me know if this looks go to merge from your end

**Comment by adam3smith** [comment]
@greebie just a brief ping -- would love to get this out and then publicize. 
One of my colleagues suggested writing a [JOSS paper](https://joss.theoj.org/)  about this -- what do you think?
Sebastian

**Comment by greebie** [comment]
My apologies -- I got buried under research stuff. It looks like you fixed all the problems I could find, so thanks. I think a JOSS paper would be great. Another option is considering a short paper for JCDL or IIPC - I think your particular use case would be of particular interest to that group.

**Pull request merged**